### PR TITLE
Added support for Snippets

### PR DESCRIPTION
**Pull request by guyzmo**
new API, that should be coming in Gitlab 8.15

cf [Gitlab-CE MR !6373](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/6373)

### CODE REVIEW

**Comment by guyzmo** [comment]
the code is still untested — and will fail up until the !6373 is accepted, but it's a start to have the new API implemented as soon as !6373 is merged and 8.15 released.

**Comment by gpocentek** [comment]
Thank you! I did a quick review and the code looks OK to me. Let's wait for the feature to land in gitlab and I'll test properly.

The unit tests fail because of a typo, shouldn't be too hard to fix :)

**Comment by guyzmo** [comment]
ok, so far so good, all tests are 👌

BTW, about APIs, you might want to have a look at upcoming changes that they're planning to do for the next major release:

* [API Changes for 9.0](https://gitlab.com/gitlab-org/gitlab-ce/issues/20070)
* [Use of swagger](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/2397)

**Comment by guyzmo** [comment]
It'd be also nice if we could throw a nice error if the API is missing (using the gitlab server's version, for example?)

**Comment by gpocentek** [comment]
I have some ideas about how we could better deal with versionning, but it would require important changes. This is getting clearer in my head by I'm afraid it will not implemented soon (lack of time basically).

I'd be happy to share if you're interested, and I'm open to suggestions :)

**Comment by guyzmo** [comment]
as an update, the API is now merged in master, and due to release in 8.15, which is supposed to be out in 3 days. 🤞

**Comment by guyzmo** [comment]
hurra 🙌 gitlab 8.15 has been released yesterday, and it's featuring [the `/snippets` endpoint](https://about.gitlab.com/2016/12/22/gitlab-8-15-released/#api-changes) through the patch I implemented.

So now we can test this PR and merge it if it's actually working fine, which it should, as I wrote the other side of the API as well 😛

**Comment by gpocentek** [comment]
@guyzmo thank you for the update and the other side of the API.

I should be able to have a look at the patches in the coming days.

File: gitlab/__init__.py [hunk 1]
```diff
@@ -484,7 +484,7 @@ def delete(self, obj, id=None, **kwargs):
             params.pop(obj.idAttr)
 
         r = self._raw_delete(url, **params)
-        raise_error_from_response(r, GitlabDeleteError)
+        raise_error_from_response(r, GitlabDeleteError, expected_code=[200, 204])
```
Comment (guyzmo) [inline gitlab/__init__.py]: cf [delete return values](https://docs.gitlab.com/ce/api/snippets.html#delete-snippet), new API spec returns 204 for deletes and return no contents.

**Comment by guyzmo** [comment]
I made a quick check with my git-repo implementation (with python-gitlab pointing to my branch) and it's working fine. I updated the tests as well, but they're failing as for some reasons [my tests are not using my fork of python-gitlab](https://travis-ci.org/guyzmo/git-repo/jobs/186391777#L395).

**Comment by gpocentek** [comment]
The test failure is related to issue #199.

**Comment by gpocentek** [comment]
Thank you!

**Comment by gpocentek** [comment]
Question: is the ``owned`` method really needed? ``list()`` will only list the current user snippets.

And I think ``all()`` should be renamed ``public()`` to avoid any confusion. What do you think?

**Comment by gpocentek** [comment]
Also the ``content()`` method is a problem when updating an object, it conflicts with the optional attribute.

**Comment by guyzmo** [comment]
> Also the content() method is a problem when updating an object, it conflicts with the optional attribute.

ah! good catch, I did not notice that… I guess `content()` should be renamed `raw()` then.

> And I think all() should be renamed public() to avoid any confusion. What do you think?

sure!

> Question: is the owned method really needed? list() will only list the current user snippets.

good catch, and yes! ☺

did you handled those when you did the merge?

**Comment by guyzmo** [comment]
(BTW, you can ping me on IRC if you use it, I'm zmo on freenode)

and happy holiday ☺

**Comment by gpocentek** [comment]
I've made the changes this morning, nothing really complicated. Thank you for you feedback.

FYI the `visibility_level` attribute doesn't seem to be exposed by the server API. Is this intended?

**Comment by guyzmo** [comment]
This is not intended and I actually did not notice, and neither did the reviewers /o\

Though I guess it's kind of ok… Someone else will have to fix that 😀

https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/6373/diffs#cc752c66542cbeeb21ca7a57c430836b865580c9_203_203

**Comment by guyzmo** [comment]
Otherwise, if you've got an idea of when you think you'll make a new release, I'd be happy to know so I can be prepared to release the new version of git-repo using it ☺

And thank you 😀

**Comment by gpocentek** [comment]
I'll probably make a new release this week to fix a few problems with gitlab 8.15.

**Pull request merged**